lib/Target/DebugInfo/EmitHGLDD.cpp

```diff
@@ -253,51 +435,360 @@ void FileEmitter::emitInstance(llvm::json::OStream &json,
+        values.push_back(value.expr);
+        if (type && type != value.type)
+          return {};
+        type = value.type;
       }
     }
-    if (portName) {
-      json.attributeObject(
-          "value", [&] { json.attribute("sig_name", portName.getValue()); });
+
+    // Emit empty arrays as 0 `bit`.
+    if (!type)
+      return {JObject({{"integer_num", 0}}), EmittedType("bit")};
+
+    type.addUnpackedDim(values.size());
+    return {JObject({
+                {"opcode", "'{"},
+                {"operands", values},
+            }),
+            type};
+  }
+
+  // Look through read inout ops.
+  if (auto readOp = dyn_cast<sv::ReadInOutOp>(op))
+    return emitExpression(readOp.getInput());
+
+  // Emit unary and binary combinational ops as their corresponding HGLDD
+  // operation.
+  StringRef unaryOpcode = TypeSwitch<Operation *, StringRef>(op)
+                              .Case<comb::ParityOp>([](auto) { return "^"; })
+                              .Default([](auto) { return ""; });
+  if (!unaryOpcode.empty() && op->getNumOperands() == 1) {
+    auto arg = emitExpression(op->getOperand(0));
+    if (!arg)
+      return {};
+    return {JObject({
+                {"opcode", unaryOpcode},
+                {"operands", JArray{arg.expr}},
+            }),
+            result.getType()};
+  }
+
+  StringRef binaryOpcode =
+      TypeSwitch<Operation *, StringRef>(op)
+          .Case<comb::AddOp>([](auto) { return "+"; })
+          .Case<comb::SubOp>([](auto) { return "-"; })
+          .Case<comb::MulOp>([](auto) { return "*"; })
+          .Case<comb::DivUOp, comb::DivSOp>([](auto) { return "/"; })
+          .Case<comb::ModUOp, comb::ModSOp>([](auto) { return "%"; })
+          .Case<comb::ShlOp>([](auto) { return "<<"; })
+          .Case<comb::ShrUOp>([](auto) { return ">>"; })
+          .Case<comb::ShrSOp>([](auto) { return ">>>"; })
+          .Case<comb::ICmpOp>([](auto cmpOp) -> StringRef {
+            switch (cmpOp.getPredicate()) {
+            case comb::ICmpPredicate::eq:
+              return "==";
+            case comb::ICmpPredicate::ne:
+              return "!=";
+            case comb::ICmpPredicate::ceq:
+              return "===";
+            case comb::ICmpPredicate::cne:
+              return "!==";
+            case comb::ICmpPredicate::weq:
+              return "==?";
+            case comb::ICmpPredicate::wne:
+              return "!=?";
+            case comb::ICmpPredicate::ult:
+            case comb::ICmpPredicate::slt:
+              return "<";
+            case comb::ICmpPredicate::ugt:
+            case comb::ICmpPredicate::sgt:
+              return ">";
+            case comb::ICmpPredicate::ule:
+            case comb::ICmpPredicate::sle:
+              return "<=";
+            case comb::ICmpPredicate::uge:
+            case comb::ICmpPredicate::sge:
+              return ">=";
+            }
+            return {};
+          })
+          .Default([](auto) { return ""; });
```

Comment:
Not blocking the PR but this duplicates ExportVerilog so it would be nice if we could declaratively define these opecodes in tablegen or somewhere

Reply:
Yeah that's a good point. I'm not entirely sure how well HGLDD will adhere to the SV operators in the future :thinking:. But you're right, as long as it stays the same, we could try to share these operators with ExportVerilog somehow.